internal/gitmetadata.go

```diff
@@ -0,0 +1,144 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+package internal
+
+import (
+	"os"
+	"reflect"
+	"runtime/debug"
+	"sync"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+const (
+	envEnabledFlag   = "DD_TRACE_GIT_METADATA_ENABLED"
+	envRepositoryURL = "DD_GIT_REPOSITORY_URL"
+	envCommitSha     = "DD_GIT_COMMIT_SHA"
+	envMainPackage   = "DD_MAIN_PACKAGE"
+	envGlobalTags    = "DD_TAGS"
+
+	tagRepositoryURL = "git.repository_url"
+	tagCommitSha     = "git.commit.sha"
+
+	traceTagRepositoryURL = "_dd.git.repository_url"
+	traceTagCommitSha     = "_dd.git.commit.sha"
+)
+
+var (
+	lock = &sync.Mutex{}
+
+	gitMetadataTags map[string]string
+)
+
+// Get git metadata from environment variables
+func getTagsFromEnv() map[string]string {
+	repositoryURL := os.Getenv(envRepositoryURL)
+	commitSha := os.Getenv(envCommitSha)
+
+	if repositoryURL == "" || commitSha == "" {
+		tags := ParseTagString(os.Getenv(envGlobalTags))
+
+		repositoryURL = tags[tagRepositoryURL]
+		commitSha = tags[tagCommitSha]
+	}
+
+	if repositoryURL == "" || commitSha == "" {
+		return nil
+	}
+
+	return map[string]string{
+		tagRepositoryURL: repositoryURL,
+		tagCommitSha:     commitSha,
+	}
+}
+
+// getTagsFrom binalry extracts git metadata from binary metadata:
+func getTagsFromBinary() map[string]string {
+	res := make(map[string]string)
+	info, ok := debug.ReadBuildInfo()
+	if !ok {
+		log.Warn("ReadBuildInfo failed, skip source code metadata extracting")
+		return res
+	}
+
+	r := reflect.ValueOf(info).Elem()
+	settings := r.FieldByName("Settings")
```

Comment:
I think it would be better to use [build constraints](https://pkg.go.dev/cmd/go#hdr-Build_constraints) than reflection for this.

You can have one file, something like `binarytags_go118.go`, which actually implements this function and just accesses `info.Settings` like normal. That file would have a `//go:build go1.18` directive. Then, you can have another file, `binarytags.go`, with a `//go:build !go1.18` directive, which implements this function but just returns an empty map.

See https://github.com/DataDog/dd-trace-go/pull/1409 for an example, where we use build constraints to use a Go 1.19+ library when it's available, and use an external library when it's not.

Reflection is a pretty big hammer. If we take the build constraint approach, we can even just drop the old code once we drop support for Go 1.17 in ~1 month.

Reply:
Super! Thanks!